// @loopback/rest/dist/rest-http-error.js
"use strict";
// Copyright IBM Corp. 2018. All Rights Reserved.
// Node module: @loopback/rest
// This file is licensed under the MIT License.
// License text available at https://opensource.org/licenses/MIT
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const http_errors_1 = __importDefault(require("http-errors"));
var RestHttpErrors;
(function (RestHttpErrors) {
    function invalidData(data, name, extraProperties) {
        const msg = `Invalid data ${JSON.stringify(data)} for parameter ${name}!`;
        return Object.assign(new http_errors_1.default.BadRequest(msg), {
            code: 'INVALID_PARAMETER_VALUE',
            parameterName: name,
        }, extraProperties);
    }
    RestHttpErrors.invalidData = invalidData;
    function unsupportedMediaType(contentType, allowedTypes = []) {
        var _a;
        const msg = ((_a = allowedTypes) === null || _a === void 0 ? void 0 : _a.length) ? `Content-type ${contentType} does not match [${allowedTypes}].`
            : `Content-type ${contentType} is not supported.`;
        return Object.assign(new http_errors_1.default.UnsupportedMediaType(msg), {
            code: 'UNSUPPORTED_MEDIA_TYPE',
            contentType: contentType,
            allowedMediaTypes: allowedTypes,
        });
    }
    RestHttpErrors.unsupportedMediaType = unsupportedMediaType;
    function missingRequired(name) {
        const msg = `Required parameter ${name} is missing!`;
        return Object.assign(new http_errors_1.default.BadRequest(msg), {
            code: 'MISSING_REQUIRED_PARAMETER',
            parameterName: name,
        });
    }
    RestHttpErrors.missingRequired = missingRequired;
    function invalidParamLocation(location) {
        const msg = `Parameters with "in: ${location}" are not supported yet.`;
        return new http_errors_1.default.NotImplemented(msg);
    }
    RestHttpErrors.invalidParamLocation = invalidParamLocation;
    RestHttpErrors.INVALID_REQUEST_BODY_MESSAGE = 'The request body is invalid. See error object `details` property for more info.';
    function invalidRequestBody() {
        return Object.assign(new http_errors_1.default.UnprocessableEntity(RestHttpErrors.INVALID_REQUEST_BODY_MESSAGE), {
            code: 'VALIDATION_FAILED',
        });
    }
    RestHttpErrors.invalidRequestBody = invalidRequestBody;
})(RestHttpErrors = exports.RestHttpErrors || (exports.RestHttpErrors = {}));
