// @loopback/rest/dist/router/redirect-route.js
"use strict";
// Copyright IBM Corp. 2019. All Rights Reserved.
// Node module: @loopback/rest
// This file is licensed under the MIT License.
// License text available at https://opensource.org/licenses/MIT
Object.defineProperty(exports, "__esModule", { value: true });
class RedirectRoute {
    constructor(sourcePath, targetLocation, statusCode = 303) {
        this.sourcePath = sourcePath;
        this.targetLocation = targetLocation;
        this.statusCode = statusCode;
        // ResolvedRoute API
        this.pathParams = [];
        this.schemas = {};
        // RouteEntry implementation
        this.verb = 'get';
        this.spec = {
            description: 'LoopBack Redirect route',
            'x-visibility': 'undocumented',
            responses: {},
        };
        this.path = sourcePath;
    }
    async invokeHandler({ response }, args) {
        response.redirect(this.statusCode, this.targetLocation);
    }
    updateBindings(requestContext) {
        // no-op
    }
    describe() {
        return `RedirectRoute from "${this.sourcePath}" to "${this.targetLocation}"`;
    }
}
exports.RedirectRoute = RedirectRoute;
