// @loopback/rest/dist/router/routing-table.js
"use strict";
// Copyright IBM Corp. 2017,2019. All Rights Reserved.
// Node module: @loopback/rest
// This file is licensed under the MIT License.
// License text available at https://opensource.org/licenses/MIT
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const assert_1 = __importDefault(require("assert"));
const debug_1 = __importDefault(require("debug"));
const http_errors_1 = __importDefault(require("http-errors"));
const util_1 = require("util");
const controller_route_1 = require("./controller-route");
const openapi_path_1 = require("./openapi-path");
const trie_router_1 = require("./trie-router");
const debug = debug_1.default('loopback:rest:routing-table');
/**
 * Routing table
 */
class RoutingTable {
    constructor(_router = new trie_router_1.TrieRouter(), _externalRoutes) {
        this._router = _router;
        this._externalRoutes = _externalRoutes;
    }
    /**
     * Register a controller as the route
     * @param spec
     * @param controllerCtor
     * @param controllerFactory
     */
    registerController(spec, controllerCtor, controllerFactory) {
        var _a;
        assert_1.default(typeof spec === 'object' && !!spec, 'API specification must be a non-null object');
        if (!spec.paths || !Object.keys(spec.paths).length) {
            return;
        }
        debug('Registering Controller with API %s', util_1.inspect(spec, { depth: null }));
        const basePath = (_a = spec.basePath, (_a !== null && _a !== void 0 ? _a : '/'));
        for (const p in spec.paths) {
            for (const verb in spec.paths[p]) {
                const opSpec = spec.paths[p][verb];
                const fullPath = RoutingTable.joinPath(basePath, p);
                const route = new controller_route_1.ControllerRoute(verb, fullPath, opSpec, controllerCtor, controllerFactory);
                this.registerRoute(route);
            }
        }
    }
    static joinPath(basePath, path) {
        const fullPath = [basePath, path]
            .join('/') // Join by /
            .replace(/(\/){2,}/g, '/') // Remove extra /
            .replace(/\/$/, '') // Remove trailing /
            .replace(/^(\/)?/, '/'); // Add leading /
        return fullPath;
    }
    /**
     * Register a route
     * @param route - A route entry
     */
    registerRoute(route) {
        // TODO(bajtos) handle the case where opSpec.parameters contains $ref
        // See https://github.com/strongloop/loopback-next/issues/435
        /* istanbul ignore if */
        if (debug.enabled) {
            debug('Registering route %s %s -> %s(%s)', route.verb.toUpperCase(), route.path, route.describe(), describeOperationParameters(route.spec));
        }
        openapi_path_1.validateApiPath(route.path);
        this._router.add(route);
    }
    describeApiPaths() {
        const paths = {};
        for (const route of this._router.list()) {
            if (route.spec['x-visibility'] === 'undocumented')
                continue;
            if (!paths[route.path]) {
                paths[route.path] = {};
            }
            paths[route.path][route.verb] = route.spec;
        }
        return paths;
    }
    /**
     * Map a request to a route
     * @param request
     */
    find(request) {
        debug('Finding route for %s %s', request.method, request.path);
        const found = this._router.find(request);
        if (found) {
            debug('Route matched: %j', found);
            return found;
        }
        if (this._externalRoutes) {
            debug('No API route found for %s %s, trying to find an external Express route', request.method, request.path);
            return this._externalRoutes.find(request);
        }
        debug('No route found for %s %s', request.method, request.path);
        throw new http_errors_1.default.NotFound(`Endpoint "${request.method} ${request.path}" not found.`);
    }
}
exports.RoutingTable = RoutingTable;
function describeOperationParameters(opSpec) {
    return (opSpec.parameters || [])
        .map(p => { var _a; return ((_a = p) === null || _a === void 0 ? void 0 : _a.name) || ''; })
        .join(', ');
}
