// @loopback/boot/dist/booters/repository.booter.js
"use strict";
// Copyright IBM Corp. 2018,2019. All Rights Reserved.
// Node module: @loopback/boot
// This file is licensed under the MIT License.
// License text available at https://opensource.org/licenses/MIT
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
const context_1 = require("@loopback/context");
const core_1 = require("@loopback/core");
const keys_1 = require("../keys");
const types_1 = require("../types");
const base_artifact_booter_1 = require("./base-artifact.booter");
/**
 * A class that extends BaseArtifactBooter to boot the 'Repository' artifact type.
 * Discovered repositories are bound using `app.repository()` which must be added
 * to an Application using the `RepositoryMixin` from `@loopback/repository`.
 *
 * Supported phases: configure, discover, load
 *
 * @param app - Application instance
 * @param projectRoot - Root of User Project relative to which all paths are resolved
 * @param bootConfig - Repository Artifact Options Object
 */
let RepositoryBooter = class RepositoryBooter extends base_artifact_booter_1.BaseArtifactBooter {
    constructor(app, projectRoot, repositoryOptions = {}) {
        super(projectRoot, 
        // Set Repository Booter Options if passed in via bootConfig
        Object.assign({}, exports.RepositoryDefaults, repositoryOptions));
        this.app = app;
        this.repositoryOptions = repositoryOptions;
    }
    /**
     * Uses super method to get a list of Artifact classes. Boot each class by
     * binding it to the application using `app.repository(repository);` if present.
     */
    async load() {
        await super.load();
        /**
         * If Repository Classes were discovered, we need to make sure RepositoryMixin
         * was used (so we have `app.repository()`) to perform the binding of a
         * Repository Class.
         */
        if (this.classes.length > 0) {
            if (!this.app.repository) {
                console.warn('app.repository() function is needed for RepositoryBooter. You can add ' +
                    'it to your Application using RepositoryMixin from @loopback/repository.');
            }
            else {
                this.classes.forEach(cls => {
                    this.app.repository(cls);
                });
            }
        }
    }
};
RepositoryBooter = __decorate([
    types_1.booter('repositories'),
    __param(0, context_1.inject(core_1.CoreBindings.APPLICATION_INSTANCE)),
    __param(1, context_1.inject(keys_1.BootBindings.PROJECT_ROOT)),
    __param(2, context_1.config()),
    __metadata("design:paramtypes", [Object, String, Object])
], RepositoryBooter);
exports.RepositoryBooter = RepositoryBooter;
/**
 * Default ArtifactOptions for RepositoryBooter.
 */
exports.RepositoryDefaults = {
    dirs: ['repositories'],
    extensions: ['.repository.js'],
    nested: true,
};
