// @loopback/context/dist/binding-inspector.js
"use strict";
// Copyright IBM Corp. 2018,2019. All Rights Reserved.
// Node module: @loopback/context
// This file is licensed under the MIT License.
// License text available at https://opensource.org/licenses/MIT
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const metadata_1 = require("@loopback/metadata");
const debug_1 = __importDefault(require("debug"));
const binding_1 = require("./binding");
const keys_1 = require("./keys");
const debug = debug_1.default('loopback:context:binding-inspector');
/**
 * Metadata key for binding metadata
 */
exports.BINDING_METADATA_KEY = metadata_1.MetadataAccessor.create('binding.metadata');
/**
 * Check if a class implements `Provider` interface
 * @param cls - A class
 */
function isProviderClass(cls) {
    return cls && typeof cls.prototype.value === 'function';
}
exports.isProviderClass = isProviderClass;
/**
 * A factory function to create a template function to bind the target class
 * as a `Provider`.
 * @param target - Target provider class
 */
function asProvider(target) {
    return function bindAsProvider(binding) {
        binding.toProvider(target).tag(keys_1.ContextTags.PROVIDER, {
            [keys_1.ContextTags.TYPE]: keys_1.ContextTags.PROVIDER,
        });
    };
}
exports.asProvider = asProvider;
/**
 * A factory function to create a template function to bind the target class
 * as a class or `Provider`.
 * @param target - Target class, which can be an implementation of `Provider`
 */
function asClassOrProvider(target) {
    // Add a template to bind to a class or provider
    return function bindAsClassOrProvider(binding) {
        if (isProviderClass(target)) {
            asProvider(target)(binding);
        }
        else {
            binding.toClass(target);
        }
    };
}
exports.asClassOrProvider = asClassOrProvider;
/**
 * Convert binding scope and tags as a template function
 * @param scopeAndTags - Binding scope and tags
 */
function asBindingTemplate(scopeAndTags) {
    return function applyBindingScopeAndTag(binding) {
        if (scopeAndTags.scope) {
            binding.inScope(scopeAndTags.scope);
        }
        if (scopeAndTags.tags) {
            if (Array.isArray(scopeAndTags.tags)) {
                binding.tag(...scopeAndTags.tags);
            }
            else {
                binding.tag(scopeAndTags.tags);
            }
        }
    };
}
exports.asBindingTemplate = asBindingTemplate;
/**
 * Get binding metadata for a class
 * @param target - The target class
 */
function getBindingMetadata(target) {
    return metadata_1.MetadataInspector.getClassMetadata(exports.BINDING_METADATA_KEY, target);
}
exports.getBindingMetadata = getBindingMetadata;
/**
 * A binding template function to delete `name` and `key` tags
 */
function removeNameAndKeyTags(binding) {
    if (binding.tagMap) {
        delete binding.tagMap.name;
        delete binding.tagMap.key;
    }
}
exports.removeNameAndKeyTags = removeNameAndKeyTags;
/**
 * Get the binding template for a class with binding metadata
 *
 * @param cls - A class with optional `@bind`
 */
function bindingTemplateFor(cls) {
    var _a, _b;
    const spec = getBindingMetadata(cls);
    debug('class %s has binding metadata', cls.name, spec);
    const templateFunctions = (_b = (_a = spec) === null || _a === void 0 ? void 0 : _a.templates, (_b !== null && _b !== void 0 ? _b : [asClassOrProvider(cls)]));
    return function applyBindingTemplatesFromMetadata(binding) {
        for (const t of templateFunctions) {
            binding.apply(t);
        }
        if (spec && spec.target !== cls) {
            // Remove name/key tags inherited from base classes
            binding.apply(removeNameAndKeyTags);
        }
    };
}
exports.bindingTemplateFor = bindingTemplateFor;
exports.DEFAULT_TYPE_NAMESPACES = {
    class: 'classes',
    provider: 'providers',
};
/**
 * Create a binding from a class with decorated metadata. The class is attached
 * to the binding as follows:
 * - `binding.toClass(cls)`: if `cls` is a plain class such as `MyController`
 * - `binding.toProvider(cls)`: it `cls` is a value provider class with a
 * prototype method `value()`
 *
 * @param cls - A class. It can be either a plain class or  a value provider class
 * @param options - Options to customize the binding key
 */
function createBindingFromClass(cls, options = {}) {
    debug('create binding from class %s with options', cls.name, options);
    try {
        const templateFn = bindingTemplateFor(cls);
        const key = buildBindingKey(cls, options);
        const binding = binding_1.Binding.bind(key).apply(templateFn);
        applyClassBindingOptions(binding, options);
        return binding;
    }
    catch (err) {
        err.message += ` (while building binding for class ${cls.name})`;
        throw err;
    }
}
exports.createBindingFromClass = createBindingFromClass;
function applyClassBindingOptions(binding, options) {
    if (options.name) {
        binding.tag({ name: options.name });
    }
    if (options.type) {
        binding.tag({ type: options.type }, options.type);
    }
    if (options.defaultScope) {
        binding.applyDefaultScope(options.defaultScope);
    }
}
/**
 * Find/infer binding key namespace for a type
 * @param type - Artifact type, such as `controller`, `datasource`, or `server`
 * @param typeNamespaces - An object mapping type names to namespaces
 */
function getNamespace(type, typeNamespaces = exports.DEFAULT_TYPE_NAMESPACES) {
    if (type in typeNamespaces) {
        return typeNamespaces[type];
    }
    else {
        // Return the plural form
        return `${type}s`;
    }
}
/**
 * Build the binding key for a class with optional binding metadata.
 * The binding key is resolved in the following steps:
 *
 * 1. Check `options.key`, if it exists, return it
 * 2. Check if the binding metadata has `key` tag, if yes, return its tag value
 * 3. Identify `namespace` and `name` to form the binding key as
 * `<namespace>.<name>`.
 *   - namespace
 *     - `options.namespace`
 *     - `namespace` tag value
 *     - Map `options.type` or `type` tag value to a namespace, for example,
 *       'controller` to 'controller'.
 *   - name
 *     - `options.name`
 *     - `name` tag value
 *     - the class name
 *
 * @param cls - A class to be bound
 * @param options - Options to customize how to build the key
 */
function buildBindingKey(cls, options = {}) {
    var _a, _b, _c, _d;
    if (options.key)
        return options.key;
    const templateFn = bindingTemplateFor(cls);
    // Create a temporary binding
    const bindingTemplate = new binding_1.Binding('template').apply(templateFn);
    // Is there a `key` tag?
    let key = bindingTemplate.tagMap[keys_1.ContextTags.KEY];
    if (key)
        return key;
    let namespace = (_a = options.namespace, (_a !== null && _a !== void 0 ? _a : bindingTemplate.tagMap[keys_1.ContextTags.NAMESPACE]));
    if (!namespace) {
        const namespaces = Object.assign({}, exports.DEFAULT_TYPE_NAMESPACES, options.typeNamespaceMapping);
        // Derive the key from type + name
        let type = (_b = options.type, (_b !== null && _b !== void 0 ? _b : bindingTemplate.tagMap[keys_1.ContextTags.TYPE]));
        if (!type) {
            type = (_c = bindingTemplate.tagNames.find(t => namespaces[t] != null), (_c !== null && _c !== void 0 ? _c : keys_1.ContextTags.CLASS));
        }
        namespace = getNamespace(type, namespaces);
    }
    const name = (_d = options.name, (_d !== null && _d !== void 0 ? _d : (bindingTemplate.tagMap[keys_1.ContextTags.NAME] || cls.name)));
    key = `${namespace}.${name}`;
    return key;
}
