// @loopback/openapi-v3/dist/decorators/index.js
"use strict";
// Copyright IBM Corp. 2018. All Rights Reserved.
// Node module: @loopback/openapi-v3
// This file is licensed under the MIT License.
// License text available at https://opensource.org/licenses/MIT
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
__export(require("./api.decorator"));
__export(require("./deprecated.decorator"));
__export(require("./operation.decorator"));
__export(require("./parameter.decorator"));
__export(require("./request-body.decorator"));
const api_decorator_1 = require("./api.decorator");
const deprecated_decorator_1 = require("./deprecated.decorator");
const operation_decorator_1 = require("./operation.decorator");
const parameter_decorator_1 = require("./parameter.decorator");
const request_body_decorator_1 = require("./request-body.decorator");
const tags_decorator_1 = require("./tags.decorator");
exports.oas = {
    api: api_decorator_1.api,
    operation: operation_decorator_1.operation,
    // methods
    get: operation_decorator_1.get,
    post: operation_decorator_1.post,
    del: operation_decorator_1.del,
    patch: operation_decorator_1.patch,
    put: operation_decorator_1.put,
    //param
    param: parameter_decorator_1.param,
    // request body
    requestBody: request_body_decorator_1.requestBody,
    // oas convenience decorators
    deprecated: deprecated_decorator_1.deprecated,
    tags: tags_decorator_1.tags,
};
