// @loopback/rest/dist/http-handler.js
"use strict";
// Copyright IBM Corp. 2017,2019. All Rights Reserved.
// Node module: @loopback/rest
// This file is licensed under the MIT License.
// License text available at https://opensource.org/licenses/MIT
Object.defineProperty(exports, "__esModule", { value: true });
const keys_1 = require("./keys");
const request_context_1 = require("./request-context");
const router_1 = require("./router");
class HttpHandler {
    constructor(_rootContext, _serverConfig, _routes = new router_1.RoutingTable()) {
        this._rootContext = _rootContext;
        this._serverConfig = _serverConfig;
        this._routes = _routes;
        this.handleRequest = (req, res) => this._handleRequest(req, res);
    }
    registerController(spec, controllerCtor, controllerFactory) {
        this._routes.registerController(spec, controllerCtor, controllerFactory);
    }
    registerRoute(route) {
        this._routes.registerRoute(route);
    }
    registerApiDefinitions(defs) {
        this._apiDefinitions = Object.assign({}, this._apiDefinitions, defs);
    }
    getApiDefinitions() {
        return this._apiDefinitions;
    }
    describeApiPaths() {
        return this._routes.describeApiPaths();
    }
    findRoute(request) {
        const route = this._routes.find(request);
        Object.assign(route.schemas, this.getApiDefinitions());
        return route;
    }
    async _handleRequest(request, response) {
        const requestContext = new request_context_1.RequestContext(request, response, this._rootContext, this._serverConfig);
        const sequence = await requestContext.get(keys_1.RestBindings.SEQUENCE);
        await sequence.handle(requestContext);
    }
}
exports.HttpHandler = HttpHandler;
